pymc3/gp/cov.py

```diff
@@ -516,3 +552,62 @@ def f(x, args):
                 args = (args,)
             return func(x, *args)
     return f
+
+
+class Coregion(Covariance):
+    R"""Covariance function for intrinsic/linear coregionalization models.
+    Adapted from GPy http://gpy.readthedocs.io/en/deploy/GPy.kern.src.html#GPy.kern.src.coregionalize.Coregionalize.
+
+    This covariance has the form:
+
+    .. math::
+
+       \mathbf{B} = \mathbf{W}\mathbf{W}^\top + \text{diag}(\kappa)
+
+    and calls must use integers associated with the index of the matrix.
+    This allows the api to remain consistent with other covariance objects:
+
+    .. math::
+
+        k(x, x') = \mathbf{B}[x, x'^\top]
+
+    Parameters
+    ----------
+    W : 2D array of shape (num_outputs, rank)
+        a low rank matrix that determines the correlations between
+        the different outputs (rows)
+    kappa : 1D array of shape (num_outputs, )
+        a vector which allows the outputs to behave independently
+    B : 2D array of shape (num_outputs, rank)
+        the total matrix, exactly one of (W, kappa) and B must be provided
+    """
+
+    def __init__(self, input_dim, W=None, kappa=None, B=None, active_dims=None):
+        super(Coregion, self).__init__(input_dim, active_dims)
+        if len(self.active_dims) != 1:
+            raise ValueError('Coregion requires exactly one dimension to be active')
+        make_B = W is not None or kappa is not None
+        if make_B and B is not None:
+            raise ValueError('Exactly one of (W, kappa) and B must be provided to Coregion')
+        if make_B:
+            self.W = tt.as_tensor_variable(W)
+            self.kappa = tt.as_tensor_variable(kappa)
+            self.B = tt.dot(self.W, self.W.T) + tt.diag(self.kappa)
+        elif B is not None:
+            self.B = tt.as_tensor_variable(B)
+        else:
+            raise ValueError('Exactly one of (W, kappa) and B must be provided to Coregion')
+
+    def full(self, X, Xs=None):
+        X, Xs = self._slice(X, Xs)
+        index = tt.cast(X, 'int32')
+        if Xs is None:
+            index2 = index.T
+        else:
+            index2 = tt.cast(Xs, 'int32').T
+        return self.B[index, index2]
```

Comment:
Requires that `X` has `input_dim=1`?  If so, should specify in docstring

Reply:
Yeah, or at least that `len(active_dims) == 1`. I added this to the notes in the docstring. Thanks!